### ISSUE TITLE
Notes on Refactoring Swingset Configuration and Startup

### ISSUE THREAD

**Issue by FUDCo** [Agoric/agoric-sdk#1744]
# Notes on Refactoring Swingset Configuration and Startup

## Vat Controller and Kernel Startup

Currently, each static vat is created during kernel startup, calling the vat's `buildRootObject` function, passing it the vat powers and vat parameters. Although vat creation is invoked from kernel code, this portion of the kernel's life cycle is orchestrated by the vat controller rather than by the kernel itself.  When replaying a vat, the initial state of the root object must be reconstructed, so its `buildRootObject` function is called _again_ at the beginning of each replay (using the same kernel and controller code that did this initially).  However, the parameters are currently supplied externally by whoever is invoking the vat controller, typically obtaining them from the command line and the environment at the time.  Consequently, in order to achieve accurate replay, special care must be taken to ensure that when the replay is started these parameters are the same as they were initially.  We would prefer that the initialization parameters instead be captured once at the birth of the swingset and saved as part of the persistent swingset state, so replay can ensure that the reconstruction of each vat's initial state is exact without having to rely on external parties to replicate their own earlier behavior.  The same reasoning applies to everything in the code bundle that defines the vat's behavior, notably including the `buildRootObject` function itself.  (A parallel argument applies when the vat is defined by a `setup` function rather than `buildRootObject`, but while this changes some details it doesn't alter the substance of the issue.)  Notably, in the case of dynamic vats the vat definitions and creation parameters _are_ recorded in the persistent kernel state and are used to replay dynamic vat creation, so the machinery to capture and use this information already exists.  However, some refactoring of the vat controller and the kernel startup code will be needed to take advantage of this.

Additionally, we note that just as vats are defined by code bundles that ought to be captured persistently, the kernel itself is defined by such a bundle. This bundle too should be saved so that replay can proceed with the same kernel that was used previously.  Note, however, that one reason for stopping and then resuming a swingset may be to upgrade the kernel code, so we should take care that the way we save and reuse the kernel bundle does not lock us into a design that precludes intentional kernel code replacement -- in this case, the primary goal is to make ordinary restarts more automatic and less susceptible to operational or administrative error.

The big picture strategy for refactoring is to split the current startup operation into two phases.  The first phase, initialization, is parameterized by the vat and kernel definitions, vat configuration options, and any other environmental information that should be fixed at startup time.  The result of this is saved as the initial contents of the kernel state database.  The second phase, execution, is parameterized only by options that control the execution process itself without affecting the semantics of the computation running in the vats (so these options might legitimately include, for example, the choice of whether and how to output runtime debugging diagnostics or performance measurement data).  The start of the execution phase also includes linkage with devices whose state, by definition, cannot be statically defined nor reconciled with the orthogonal persistence model used by vats and the kernel.  However, all information that needs to carry over from the initialization phase to the execution phase must be captured in the kernel state database.

In this new model, we unify initial execution with replay execution.  In essence, an initial execution _is_ a replay, one that proceeds from a state database persisted at crank zero, in which all vats are in their initial states after return from `buildRootObject` and the kernel run queue consists solely of the `bootstrap` message that will begin actual swingset execution when delivered to the bootstrap vat during crank one.

### Current Schema

Every vat, static or dynamic, is identified by a vatID.  These are currently allocated sequentially via a counter stored as:

`vat.nextID = $NN`

The current kernel schema represents static vats with:

`vat.names = JSON([names..])`
`vat.name.$NAME = $vatID = v$NN`

Every static vat has a name.  For each static vat, `vat.name.$NAME` maps to the vatID of the vat with the corresponding name.  A complete bidirectional mapping between static vat names and vatIDs can be constructed by iterating over all the `vat.name.*` entries.  In addition, the value of `vat.names` is a complete list of all current static vat names.  This latter is not strictly needed but is provided as a convenience; it is currently only used in the generation of the `bootstrap` message.

Dynamic vats are currently all unnamed.  The complete list of the vatIDs of existing dynamic vats is stored as:

`vat.dynamicIDs = JSON([vatIDs..])`

In addition, we store the source and configuration options of each dynamic vat with:

`v$NN.source = JSON({ bundle }) or JSON({ bundleName })`
`v$NN.options = JSON`

### Proposed Schema Changes

The minimal change we need is to add the `v$NN.source` and `v$NN.options` values for static vats.  I believe this is entirely sufficient for our immediate needs.

However, given that we are revisiting the schema anyway, there may be a few improvements we might wish to incorporate while we're at it.

I think the `vat.names` list can be eliminated, since it contains no information that is not already available from the set of `vat.name.$NAME` values.

In some of our in-person discussions we've tossed around the idea of supporting names for dynamic vats, but upon reflection I don't think this is a requirement or even useful.  Currently, access to the vat name to vatID mapping is only available within the kernel itself (which makes sense since one of liveslots' jobs is insulating vat code from ever seeing or using vatIDs anyway).  Vat names currently get used in four ways:

- to generate the parameters to the `bootstrap` message, so the bootstrap vat can have stable labels for the other static vats without getting tangled up in how the kernel assigns vatIDs; this construction happens once at startup time.

- so that the kernel itself can have similar stability for referring to the various static internal vats that are entangled with swingset operation (such as the vat admin vat).

- to target messages sent by the host process using the vat controller's `queueToVatExport` method; these messages are for testing and diagnostic purposes and not considered part of the normal vat API.

- to look up static vats in unit tests that have no bootstrap vat (using a special controller APIs that are only available for testing).

None of these use cases manifest a need for kernel-managed names for dynamic vats, nor is it obvious how or to whom such a facility might be exposed if it existed.  Consequently, I don't think any changes are needed to the vat naming schema.

The `vat.dynamicIDs` value exists so that the complete set of vats can be enumerated when starting a replay.  It need only list the dynamic vats because the static vats can already be determined from the `vat.name.$NAME` entries. This mechanism works as is, but since the entire dynamic vat list is a single key-value entry, updating it as vats come and go may become expensive if the number of running dynamic vats grows large.  One way to deal with this would be to go ahead and, despite the argument presented in the previous paragraphs, give names to dynamic vats and allow the `vat.name.$NAME` values serve for identifying all vats.  Presumably the names would be synthetic and automatically generated, in order to avoid duplications that would result in key collisions. This feels a bit kludgy to me.  I think I'd rather introduce a new set of keys of the form `vat.dynamic.v$NN`.  Since these keys' sole reason for existence is to be enumerated, the associated value is somewhat arbitrary.  It could be left as the empty string, or set to a fixed value (perhaps the string `true` as we do for booleans), or set to the vatID redundantly, or perhaps it could contain some kind of status indicator.  For now I'd go with the empty string.

## API Changes

Currently, `controller.js` provides `buildVatController(config, argv, runtimeOptions)` which returns a vat controller object with methods like `run`, `step`, and so on, plus methods that interrogate and poke at the kernel for testing purposes.  The `config` and `argv` parameters are initialization parameters that, per the above discussion, should be persisted.  The `runtimeOptions` parameter contains genuine runtime options, but in the current organization of the world this includes the host storage object (i.e., the thing that holds the kernel state database) which is also something needed at initialization time.

Also, the name `buildVatController` is not ideal.  It does produce a controller object, but it's a controller for the kernel or the swingset (depending on your point of view).  It also, of course, tangles up initialization in the ways discussed above.

I propose splitting this into two functions:

### `buildSwingset(config, argv, storage)`

- `config` and `argv` are the same as the corresponding paramaters to `buildVatController`

- `storage` is an optional storage object that will default to a fresh in-memory store usable only for testing.  (This storage object is the same as what was passed as the optional `hostStorage` property of the `runtimeOptions` parameter in the previous `buildVatController` call.)

The return value will be the storage object that was actually populated, either `storage` or the fresh in-memory store that `buildSwingset` constructed.

If the `storage` object provided was a persistent store, after `buildSwingset` returns the host could exit and the contents of the store used in a later, separate execution of the same or a different host program.

### `makeSwingsetController(storage, runtimeOptions)`

- `storage` is a previously populated kernel data store either as initially generated by a prior call to `buildSwingset` or as left over after an earlier swingset execution

- `runtimeOptions` is an options bag essentially identical to that now accepted by `buildVatController` except that the kernel data store has been extracted as a separate parameter.

The return value will be a controller object that is the same as is now returned by `buildVatController`.

## Storage

The swing-store call `makeSwingStore` (for each of the various species of stores we support) takes a `dirPath` parameter that indicates the directory into which the store implementation should place any files it creates for saving data. This pathname is essentially the designator for a persistent storage instance. I propose that we formalize the de facto semantics of this pathname as a designator and to make explicit the idea that this directory is for the exclusive use of a particular swingset instance to store its associated persistence data but without reference to which storage mechanism is being used for which data.  This way the implementation can use it as a reliable place to save other objects that should be associated with the persistent state but which might not be captured by the database per se.  The initial intent of this change is to permit this directory to be used to store the kernel code bundle that `buildSwingset` generates, though we anticipate the persistence model becoming more complicated as the software evolves and so other uses for this directory may also emerge (for example, once we acquire the ability to make heap snapshots, these will need to be kept somewhere).  In particular, between reserving this directory and the saving of vat code bundles inside the kernel database, it means we should be able to substitute internal-to-the-swingset mechanisms for much of the brittle and ad hoc temp file manipulation that Cosmic Swingset and ag-solo now have to enage in.

## Implementation Notes

Along with the proposed split of the controller API, it will probably make sense to also split `kernel.js` into two corresponding pieces, one for initialization and setup and one for runtime.  Since `kernel.js` is already the largest file in the kernel, it's already ripe for some restructuring anyway.  This refactoring can be done incrementally as we proceed with the controller work, letting the things we discover long the way teach us what stuff should go where.

## Longer Term Considerations

In order to support evolution of vat code in long lived swingsets, we will at some point wish to store multiple versions of a vat's code bundle, tagged according to the cranks for which they are relevant.  This would entail changing the schema for saving a vat's source to be something like:

`v$NN.source.$VERS = JSON({ bundle })`

accompanied by something like:

`v$NN.version.$VERS = $CRANKNUMBER`

to indicate the crank at which the indicated vat version became active.

Implementing a version switchover mechanism is out of scope for this particular refactoring excercise, but it is worth asking if we should change the schema now in anticipation of this change, or whether it would be better to wait until the vat upgrade story is more fully baked.

**Comment by FUDCo** [comment]
@warner

**Comment by warner** [comment]
Wow, thanks for the detailed writeup. Awesome.

A few notes:

## Names

What do you think about calling the first function `initializeSwingset` instead of `buildSwingset`, under the theory that the verb "initialize" is more obviously "call only once" than "build"?

## Devices

I think we need to pass a distinct `devices` or `deviceConfig` argument into `makeSwingsetController`, in addition to the `config.devices` property that goes into the initialization call. The init call gets device names and source specifications. The controller-building call gets device names and endowments (which contain objects/functions that cannot be saved in the DB and so cannot appear until the real runtime-hosting non-initialize-time process is running)

## Store Directory Layout

I'm mostly positive on the "store other stuff in the store directory" idea, although there's a question of what boundaries we establish.

The initial notion of the swing-store `storeDir` was that it's for the exclusive use of the swing-store implementation, so whether it contains one file or a dozen, all the caller needs to know is that if it provides the same `storeDir` to the same implementation, it will get back the same data. We relaxed that a bit when we were having LMDB problems on WSL: the init-time code figures out what DB type is best, creates it, and then the run-time code examines the directory with `getBestSwingStore()` to discover what type was picked earlier. That changes the contract to "provide the same `storeDir`", without requiring the run-time code know the exact type.

But in both cases, the swing-store code doesn't have any restrictions on the filenames it uses, and doesn't need to worry about collisions with other files. Likewise the kernel doesn't need to know what filenames the swingstore uses, since it never puts files in the same directory.

But of course the downside is we need a higher-level "basedir" notion, which is for the exclusive use of the swingset instance, and *includes* a swing-store subdirectory (which itself is for the exclusive use of the swing-store). Even if the swing-store only needs one or two files. This notion is somewhat exposed already, e.g. cosmic-swingset creates a directory for each solo/chain instance that includes both a `vats` subdirectory (which will go away once this refactoring is done) and a `swingset-kernel-state` directory.

So I think I agree that we should redefine the swingstore directory to contain all kernel state, not just the parts that go into the DB, but with a few considerations:

* the swingstores need to publish the filenames they use in the docs, so that other code can share the directory without fear of name collisions
  * maybe the swingstore API should change to provide a filename (and derive e.g. `lock.mdb` from that filename) rather than a directory, I'm not sure
* we need to keep careful track of atomicity/transactionality properties of any additional data we put in this directory: everything in the DB will be atomically committed, but random other files will either be written+closed before the DB transaction (and thus restarts might see "data from the future"), or after (so restarts might observe the data being missing). I think we'll be ok as long as we stick to a rule that the DB is canonical. So we could e.g. store large bundle strings or XS vat checkpoints in a file, indexed by a hash of the contents, and record the hash in the DB. At runtime, we'd never ever look on disk for a bundle unless we'd first seen evidence in the DB that it ought to exist.

**Comment by FUDCo** [comment]
`initializeSwingset` seems like a better name.  Let's go with that.

A devices parameter for `makeSwingsetController` makes sense and was sort of implied by but never actually followed through on in my writeup.  For now I'll add a parameter, but likely there'll still be a bit of handwaving about the specifics until we've figured them out; I think this piece is going to be a discovered design rather than something we can invent in our heads de novo.

I think the store directory has (and will have) few enough things in it that we can manage the namespace directly without having to add levels of structure, but documenting what's going in there makes a lot of sense.  Right now the only other thing that would go in there is the kernel bundle, which will get written before the state db is even created, so I think we're fine w.r.t atomicity et al at this point.  In the future, my expectation is that if things get more complicated (e.g., if we use separate stores for streamed data vs. random access data), it will be under the control of a more sophisticated storage component that will be responsible for ensuring that good ACID hygiene is obeyed.  But I can easily imagine something like a database whose implementation uses multiple files (LMDB already uses 2) and this dir makes sense as a place to put them.

**Comment by FUDCo** [comment]
Closed by #1814

**Comment by FUDCo** [comment]
Closed by #1814
